engines/registry/app/models/access_scope.rb

```diff
@@ -0,0 +1,92 @@
+require 'yaml'
+
+AUTHORIZED_ACTION = ['pull'].freeze
+# this is analogous to auth.Access in golang code.
+class Registry::AccessScope
+  attr_accessor :type,
+                :class,
+                :namespace,
+                :image,
+                :actions
+
+  # Parses a String into an authorization scope.
+  #
+  # scope - a String containing an authorization scope in the following format:
+  #         `<type>[(<class>)]:<namespace>/<image>:<actions>`
+  #         `repository(special):name:pull`
+  #         `registry:catalog:*`
+  #
+  # Returns a new Scope.
+  def self.parse(scope)
+    raise_on_invalid_scope(scope)
+    type, name, actions = scope.split(':')
+    _, type, klass = /(\w+)\(?(\w+)?\)?/.match(type).to_a
+    actions = actions.split(',')
+    new(type: type, klass: klass, name: name, actions: actions)
+  end
+
+  def initialize(type:, name:, actions:, klass: nil)
+    @type = type
+    @klass = klass
+    @namespace, _, @image = name.rpartition('/').map(&:strip).map(&:presence)
+    @actions = actions
+  end
+
+  def full_type
+    return "#{@type}(#{@klass})" if @klass.present?
+
+    @type
+  end
+
+  def to_s
+    "#{full_type}:#{name}:#{@actions.join(',')}"
+  end
+
+  def granted(client: nil)
+    aa = authorized_actions(client)
+    Rails.logger.info "Granted actions for user '#{client&.account || '<anonymous>'}': #{aa}"
+    {
+      'type' => @type,
+      'class' => @klass,
+      'name' => name,
+      'actions' => aa
+    }
+  end
+
+  def name
+    [namespace, image].map(&:presence).compact.join('/')
+  end
+
+  def authorized_actions(client = nil)
+    if @namespace.nil?
+      @image == 'catalog' ? @actions : AUTHORIZED_ACTION
+    else
+      allowed_paths(client.systems.first)
+      if @allowed_paths.any? { |allowed_path| File.fnmatch(@namespace + '*', allowed_path) }
+        @actions & AUTHORIZED_ACTION
+      else
+        []
+      end
+    end
+  end
+
+  def self.raise_on_invalid_scope(scope)
+    # if nothing is passed, return
+    raise Registry::Exceptions::InvalidScope.new('Empty scope') if scope.blank?
+    # if scope is malformed, return
+    raise Registry::Exceptions::InvalidScope.new('Invalid scope format') unless scope.split(':').size == 3
+    raise Registry::Exceptions::InvalidScope.new('Invalid scope format') unless %r{^[a-z0-9\-_/:*(),.]+$}i.match?(scope)
+  end
+
+  def allowed_paths(system = nil)
+    repo_list = RegistryCatalogService.new.repos(reload: false, system: system)
+    access_policies_yml = YAML.safe_load(
+      File.read(Rails.application.config.access_policies)
+    )
+    active_products = system.activations.includes(:product).pluck(:product_class)
+
+    allowed_products = (active_products & access_policies_yml.keys) + ['free']
+    allowed_glob_paths = access_policies_yml.values_at(*allowed_products).flatten
+    @allowed_paths = repo_list.select { |repo| repo if allowed_glob_paths.any? { |glob_path| File.fnmatch(glob_path, repo) } }
```

Comment:
`File.fnmatch()` does not handle * and ** the same way as we defined it in SCC. So this does not give the right results!
For example `File.fnmatch('*')` matches *all* repos in the registry and not only those in the "root" directory as SCC does.

Reply:
mmmm, I had run some tests and the results seem OK ?

```
> podman search registry-ec2.susecloud.net/bci
NAME                                                 DESCRIPTION
registry-ec2.susecloud.net/opensuse/bci/bci-busybox  
registry-ec2.susecloud.net/opensuse/bci/ruby         
> podman search registry-ec2.susecloud.net/tool
NAME                                                                                                         DESCRIPTION
registry-ec2.susecloud.net/opensuse/leap/micro/5.2/images/totest/containers/opensuse/leap-micro/5.2/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.3/images/totest/containers/opensuse/leap-micro/5.3/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.4/images/totest/containers/opensuse/leap-micro/5.4/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/images/opensuse/leap-micro/5.5/toolbox             
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/totest/containers/opensuse/leap-micro/5.5/toolbox  
```
and no term/pattern returns _all_ free repos 
```
> podman search registry-ec2.susecloud.net/
NAME                                                                                                         DESCRIPTION
registry-ec2.susecloud.net/home/marcus.schaefer/delta_containers/containers_tw/fleet                         
registry-ec2.susecloud.net/home/marcus.schaefer/delta_containers/containers_tw/joe                           
registry-ec2.susecloud.net/opensuse/bci/bci-busybox                                                          
registry-ec2.susecloud.net/opensuse/bci/ruby                                                                 
registry-ec2.susecloud.net/opensuse/leap/micro/5.2/images/totest/containers/opensuse/leap-micro/5.2/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.3/images/totest/containers/opensuse/leap-micro/5.3/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.4/images/totest/containers/opensuse/leap-micro/5.4/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/images/opensuse/leap-micro/5.5/toolbox             
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/totest/containers/opensuse/leap-micro/5.5/toolbox  
> 
```

are those results not OK ?

Reply:
What @cwh42 meant is the meaning of * and ** in access policies: 

'*' applies to all images in the current path. For example, the policy `suse/*` applies to all images in the `suse/` directory, like `suse/389-ds`. But not to `suse/ltss/sle15.3/bci-base` for example. 

![image](https://github.com/SUSE/rmt/assets/582520/deafc702-5373-402f-a6f6-dc83efbeecaf)

'**' applies to all images and images in subdirectories of the path. Example for `suse/sles/**`: 

![image](https://github.com/SUSE/rmt/assets/582520/7b2101ad-a445-45ef-b4a4-2891eae17983)

Reply:
Updated